keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2164 @@
+either at the same time or one after the other. Such persistent ResourceClaims
+get referenced in the pod spec by name. When a PodTemplateSpec in an app
+controller spec references a ResourceClaim by name, all pods created by that
+controller also use that name and thus share the resources allocated for that
+ResourceClaim.
+
+But often, each Pod is meant to have exclusive access to its own ResourceClaim
+instance instead. To support such ephemeral resources without having to modify
+all controllers that create Pods, an entry in the new PodSpec.ResourceClaims
+list can also be a ResourceClaimTemplate. When a Pod gets created, such a
+template will be used to create a normal ResourceClaim with the Pod as owner
+with an
+[OwnerReference](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#OwnerReference)),
+and then the normal allocation of the resource takes place. Once the pod got
+deleted, the Kubernetes garbage collector will also delete the
+ResourceClaim.
+
+The difference between persistent and ephemeral resources for kube-scheduler
+and kubelet is that the name of the ResourceClaim needs to be determined
+differently: the name of an ephemeral ResourceClaim is `<pod>-<resource name in
+the pod spec>`. Ownership must be checked to detect accidental conflicts with
+persistent ResourceClaims or previous incarnations of the same ephemeral
+resource. This is the same approach that was chosen for [generic ephemeral
+volumes](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/1698-generic-ephemeral-volumes/README.md).
+For a resource driver there is no difference.
+
+### Coordinating resource allocation through the scheduler
+
+For immediate allocation, scheduling Pods is simple because the
+resource is already allocated and determines the nodes on which the
+Pod may run. The downside is that pod scheduling is less flexible.
+
+For delayed allocation, a node is selected tentatively by the scheduler
+and driver(s) try to allocate their resources for that node. If that
+succeeds, the Pod can get scheduled. If it fails, the scheduler must
+determine whether some other node fits the requirements and if so,
+request allocation again. If no node fits because some resources were
+already allocated for a node and are only usable there, then those
+resources must be released and then get allocated elsewhere.
+
+This is a summary of the necessary [kube-scheduler changes](#kube-scheduler) in
+pseudo-code:
+
+```
+while <pod needs to be scheduled> {
+  <choose a node, considering potential availability for those resources
+   which are not allocated yet and the hard constraints for those which are>
+  if <no node fits the pod> {
+    if <at least one resource
+            is allocated and unused,
+            uses delayed allocation, and
+            was not available on a node> {
+      <randomly pick one of those resources and tell resource driver to free it>
```

Comment:
> I was fearing that this could lead to deadlocks where two or more pods continuously deallocate each others ResourceClaims.

They would compete with each other, but there would be no deadlock. It would become a timing question whether the system settles into a stable state (= one pod running), though.

> But if a ResourceClaim is allocated, it also sets ReservedFor... unless it's immediate allocation.

I don't think it would be needed to avoid a deadlock, but I agree that it helps because ReservedFor avoids threshing.

We just need to account for "reserved for the current, pending pod" in the if check: "and unused" must become "and unused or reserved for the current pod".

"tell resource driver to free it" must be "tell resource driver to free it by setting `Deallocate` and removing the pod from `ReservedFor` (if present there)".

I've changed this.

> The only scenario where a ResourceClaim would be allocated but unused is if a Pod that doesn't own it finishes running, right?

Yes.

> Why wouldn't the driver deallocate if it reaches that state?

Because of the use case of reusing a resource that is expensive to initialize: we want to keep that allocated and thus initialized according to the parameters to amortize the cost of initialization over multiple pods.

Reply:
Regarding "free" vs. "deallocate" - I consider those synonymous, so I ended up not using one or the other consistently. I've replaced "free" with "deallocate" now, to be consistent.

The additional benefit is that "deallocation" is a proper noun, which sometimes reads better.

Reply:
> Because of the use case of reusing a resource that is expensive to initialize: we want to keep that allocated and thus initialized according to the parameters to amortize the cost of initialization over multiple pods.

That sounds like a security risk, but I'll let sig-security comment on that.

Reply:
The same claim can only get reused inside the same namespace. That's typically the boundary between different users (or group of users). Whether reusing the same claim then makes sense becomes a usability and app deployment question, but shouldn't be a security risk.